### DISCUSSION TITLE
Custom FreeRDP client does not work with touch input

### DISCUSSION THREAD

**Discussion by Saddingt0n** [FreeRDP/FreeRDP#8533] [Q&A]
I managed to create my own client based on Sample client. I use this client to remote to several desktop PC with the following specs:
- OS: Ubuntu 18/20
- CPU: Intel i7-7700
- RAM: 8GB/16GB
- RDP server: xrdp (http://xrdp.org/)

Stream and control work fine. But I ran into problem with a Linux based machine (let's call this machine B). It uses touch input and RDP backend in weston. Streaming stills work but touch control doesn't.

I use the `freerdp_input_send_mouse_event` function as described in this document:
https://github.com/awakecoding/FreeRDP-Manuals/blob/master/Developer/FreeRDP-Developer-Manual.markdown#mouse-event

I think the problem is that those Ubuntu desktop machines use mouse click as input and machine B uses touch as input so the method for sending input is different? If you need any more information to diagnose the problem, please let me know.

**Comment by Saddingt0n** [comment]
I forgot to mention this. I used xfreerdp to remote to machine B and I can use mouse click for touch input just fine.

**Comment by Saddingt0n** [comment]
An update to this Q&A.

I see that with xfreerdp client, every time the cursor moves, it calls the `freerdp_input_send_mouse_event(input, PTR_FLAGS_MOVE, x, y);` method.
I also called that method before calling `freerdp_input_send_mouse_event(input, <appropriate flags here>, x, y);` for touch down/up actions but it still doesn't work so this is not the problem.

**Comment by akallabeth** [comment]
@Saddingt0n you seem to mix up `mouse` input with `touch` input.

1. You need a setup where you can use touch input on your local `X11` system
2. Your server needs support for touch input
1. touch input support is initialized with `xfreerdp +multitouch` and handled by the [MS-RDPEI](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpei/72a8cb65-7f6c-407c-a21a-3d970721fed0) channel.

So, while some touch panels emulate a mouse, touch input (as in more than one finger) requires that channel, see `xf_input_touch_remote` in `xf_input.c`

**Reply by Saddingt0n** [reply]
> @Saddingt0n you seem to mix up `mouse` input with `touch` input.
> 
> 1. You need a setup where you can use touch input on your local `X11` system
> 2. Your server needs support for touch input
> 3. touch input support is initialized with `xfreerdp +multitouch` and handled by the [MS-RDPEI](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpei/72a8cb65-7f6c-407c-a21a-3d970721fed0) channel.
> 
> So, while some touch panels emulate a mouse, touch input (as in more than one finger) requires that channel, see `xf_input_touch_remote` in `xf_input.c`

Thank you @akallabeth! 

I built `xfreerdp` with some `printf` statements in `xf_input_touch_remote` but it seems that the function is not executed once.

It seems that I am confused between mouse input and touch input.
My apology if I didn't provide enough info. I will describe it in more details here:
- I use machine A (an Ubuntu machine), with mouse and keyboards for input.
- I use xfreerdp to remote to machine B (a Linux based machine) that uses finger touch for input. Here's my command: xfreerdp <ip here>. No "+multitouch" option.
- When FreeRDP GUI window opens, I can use mouse clicks to interact with machine B just like if 

Now with my custom client:
- I use this function for sending input: `freerdp_input_send_mouse_event`
- I use the same machine A as above.
- I can use my client connecting to another Ubuntu machine C (also with mouse and keyboards for input). Here's my command: myclient /v:<ip here> /u:<username here> /p:<password here>. I can interact with machine C.
- But my client does not work with machine B which uses finger touch for input.

Do you have any ideas as to why this is the case?

**Reply by akallabeth** [reply]
@Saddingt0n ok, sounds like you are asking the wrong question here. It does not matter which kind of input your machine uses locally.
For RDP the following is of interest:
1. Does my remote OS support multitouch
1. Does my RDP server support multitouch
2. Does my RDP client support multitouch
3. Does my local OS/setup support multitouch

so, you could easily create a multitouch input connection to a remote which just uses keyboard/mouse locally or a normal keyboard/mouse connection to a system that uses multitouch locally.

As you described, that works with your first setup just fine.
Since I don´t know what you do in your client I can´t tell what your problem is, but it is not really multitouch related ;)

**Reply by Saddingt0n** [reply]
Thank you @akallabeth! 

My client has a server that takes touch input request from an outside source. For each input request (consisting of x, y coordinates and mouse click type), here's what happens:
```
if (request == left_mouse_click_down) {
    flags = PTR_FLAGS_DOWN | PTR_FLAGS_BUTTON1;
}

if (request == left_mouse_click_up) {
    flags = PTR_FLAGS_BUTTON1;;
}

bool ret = freerdp_input_send_mouse_event(input, flags, x, y); 
```
Is there anything I'm missing here?

**Reply by Saddingt0n** [reply]
> Thank you @akallabeth!
> 
> My client has a server that takes touch input request from an outside source. For each input request (consisting of x, y coordinates and mouse click type), here's what happens:
> 
> ```
> if (request == left_mouse_click_down) {
>     flags = PTR_FLAGS_DOWN | PTR_FLAGS_BUTTON1;
> }
> 
> if (request == left_mouse_click_up) {
>     flags = PTR_FLAGS_BUTTON1;;
> }
> 
> bool ret = freerdp_input_send_mouse_event(input, flags, x, y); 
> ```
> 
> Is there anything I'm missing here?

@akallabeth, I want to add a bit of info. The version of wayland on machine B is 8.0. If I'm not mistaken, the latest stable version is 9.0. 
So maybe there are some differences between the two versions and my code does not compensate for those changes?

**Reply by Saddingt0n** [reply]
@akallabeth, could you take a look at my code above when you have the time? I would appreciate it a lot, thank you!
